### DISCUSSION TITLE
Dealing with unequal length and irregular time series in time series classification, regression, forecasting, clustering, etc

### DISCUSSION THREAD

**Discussion by fkiraly** [alan-turing-institute/sktime#4300] [Q&A]
As this is a frequently asked question, I'm summarizing an answer to this.
(mid-term, we probably want to put this in the tutorials)

There are two main ways to deal with unequal length or irregularly spaced time series:

* search for an estimator of the right type, e.g., classifier, clusterer, that can deal with irregularly spaced time series. The tag in question is `capability:unequal_length` - despite the name, afaik current estimators of this type can deal with *both* unequal length and irregular spacing. To search, use `registry.all_estimators` (and see docstring).
* pipelining with a transformer that renders series equally spaced. The tag for this is `capability:unequal_length:removes`, search with `registry.all_estimators`. To pipeline, use `pipeline.make_pipeline` (and see docstring). Eamples would be `TimeBinAggregate` from `series.binning`, or `TSInterpolator` from `panel.interpolate`.

If the above does not work as implied, please raise a bug report here:
https://github.com/sktime/sktime/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBUG%5D

**Comment by philip-ndikum** [comment]
Also have the same question - any update on time series forecasters

**Reply by fkiraly** [reply]
@anonymouspapi, can you kindly explain what you mean - what precisely is your question?

**Comment by jonaslandsgesell** [comment]
Is it or will it be possible to make use of numpy masked arrays https://numpy.org/doc/stable/reference/maskedarray.generic.html to pass time series of unequal length to sktime?

I imagine to pass a numpy (masked) array X with X.shape = [num_time_series, num_features, max_length_time_series]
Now not all time series would have to be of length max_lenghr_time_series. The time series which are shorter would be filled with a mask value (fill value) so that the subsequent pipelines do not make inappropriate use of these fill values.

Alternatively: is there a conversion function for sich numpy masked arrays to a suitable pandas data frame (if that exists)?

**Reply by fkiraly** [reply]
@jonaslandsgesell, apologies, did not see this.

I was not aware of masked arrays, I think this is an excellent suggestion!

For conversion, you would use `datatypes.convert` or `convert_to`, not sure though whether that works for masked arrays.

**Reply by muratti18462** [reply]
@jonaslandsgesell Did this solution worked for your case ? and should you also use some sort of resampling method before do this ? 

I am currently working something similar , [https://stackoverflow.com/questions/79580999/unequal-time-interval-and-unequal-time-series-modelling](url)

**Comment by travis-leith** [comment]
@fkiraly any progress on the mid term tutorial idea you mentioned? Is the advice in your OP still valid?
